### PR TITLE
operator bootstrap admin handling

### LINKED ISSUE #30004: Bootstrapping an admin user or service account using the Operator
### Description

Based on the design as specified in https://github.com/keycloak/keycloak/issues/9829.

Add a dedicated field to the Keycloak CR for bootstrapping the admin user account and admin service account. The Operator will set the following env vars on the pods: `KEYCLOAK_BOOTSTRAP_ADMIN_USER`, `KEYCLOAK_BOOTSTRAP_ADMIN_PASSWORD`, `KEYCLOAK_BOOTSTRAP_ADMIN_CLIENT_ID`, `KEYCLOAK_BOOTSTRAP_ADMIN_CLIENT_SECRET`, `KEYCLOAK_BOOTSTRAP_ADMIN_EXPIRATION`. That means it will support bootstrapping only at startup time.

**Example spec:**
```
kind: Keycloak
spec:
  bootstrap-admin:
    expiration: 260
    user:
      nameSecret: 
        name: keycloak-admin-secret
        key: username
      passwordSecret:
        name: keycloak-admin-secret
        key: password
    service:
      clientIdSecret: 
        name: keycloak-admin-secret
        key: client-id
      clientSecret:
        name: keycloak-admin-secret
        key: secret

```

* The content of Secrets referenced by `nameSecret` and `passwordSecret` can be provided by the user, or the Operator will populate them with defaults.
* If the Secret or the Key referenced by `nameSecret` don’t exist, the default `temp-admin` is used and is written back to the Secret.
* If the Secret or the Key referenced by `passwordSecret` don’t exist, a password is generated and is written back to the Secret.
* The content of Secrets referenced by `clientIdSecret` and `clientSecret` must be provided by the user. `clientIdSecret` defaults to `temp-admin-service`.

### PR DESCRIPTION
**Pull request by shawkins**
This is layered on the other bootstrap changes. After looking at the implementation possiblities a little more, I think it makes more sense to manage things at a secret level - giving us 2 secrets to worry about instead of 4 - and just assume the keys by convention.

To minimize the number of api server interactions per reconciliation we also won't attempt to write back values into user provided secrets - if the user specifies the secret, then we expect it to exist and the fields be present. We can of course add optional handling if that seems warrented.

@vmuzikar WDYT?

### CODE REVIEW

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/BootstrapAdminSpec.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/BootstrapAdminSpec.java:35]:
```suggestion
    @JsonPropertyDescription("Configures the bootstrap admin service account")
```

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/BootstrapAdminSpec.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/BootstrapAdminSpec.java:8]: Should we add some desc about the key names?

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/BootstrapAdminSpec.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/BootstrapAdminSpec.java:20]: Should we add some desc about the key names?

**Review by vmuzikar — COMMENTED** [review]
@shawkins Thank you for the Operator changes. Overall looks good, added just some minor comments. We'll also need some tests.

**Comment by vmuzikar** [comment]
> To minimize the number of api server interactions per reconciliation we also won't attempt to write back values into user provided secrets - if the user specifies the secret, then we expect it to exist and the fields be present.

This works for me. We just populate the default Secret.

We probably should also add some docs and reflect it there.

**Comment by shawkins** [comment]
> We probably should also add some docs and reflect it there.

Added some docs to the advanced guide, see if that works for you.

File: docs/guides/operator/advanced-configuration.adoc
Comment (vmuzikar) [inline docs/guides/operator/advanced-configuration.adoc:313]:
```suggestion
When you create a new instance the Keycloak CR spec.bootstrapAdmin stanza may be used to configure the bootstrap user and/or service account. If you do not specify anything for the spec.bootstrapAdmin, the operator will create a Secret named "metadata.name"-initial-admin with a username temp-admin and a generated password. If you specify a Secret name for bootstrap admin user, then the Secret will need to contain `username` and `password` key value pairs. If you specify a Secret name for bootstrap admin service account, then the Secret will need to contain `client-id` and `client-secret` key value pairs.
```

File: docs/guides/operator/advanced-configuration.adoc
Comment (vmuzikar) [inline docs/guides/operator/advanced-configuration.adoc:316]: It might be also good to add a link here to the Admin Bootstrapping guide (once it's merged).

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/BootstrapAdminSpec.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/BootstrapAdminSpec.java:8]:
```suggestion
        @JsonPropertyDescription("Name of the secret that contains username and password keys")
```

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/BootstrapAdminSpec.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/BootstrapAdminSpec.java:21]:
```suggestion
        @JsonPropertyDescription("Name of the Secret that contains client-id and client-secret keys")
```

**Review by vmuzikar — COMMENTED** [review]
@shawkins Added some last suggestions, nothing really blocking.

File: docs/guides/operator/advanced-configuration.adoc
Comment (shawkins) [reply docs/guides/operator/advanced-configuration.adoc:316]: Ok, we can wait for #31039 to be merged first

**Comment by shawkins** [comment]
The strategy of modifying the Keycloak spec in the controller does not work well here - the dependents run before the controller, so we end up with a race condition as to whether a deployment is attempted without a bootstrap admin secret. Rather than trying to customize the operator workflow, the simplest thing is just to treat a non-specified user secret as the default.

**Review by vmuzikar — APPROVED** [review]
The changes still look good to me. I'd like to merge this rather sooner than later. We can add the docs link as a follow up. @shawkins WDYT?

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java
Comment (shawkins) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java:108]: @vmuzikar I don't think it's a bug, just how the operator sdk works by default.

**Comment by shawkins** [comment]
> The changes still look good to me. I'd like to merge this rather sooner than later. We can add the docs link as a follow up. @shawkins WDYT?

Sounds good.

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java:108]: Then how are we supposed to update spec?

Comment (shawkins) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java:108]: We can update it here, but not in a way that will cause a race condition. The pr initially didn't assume a default in the deployment dependent logic which lead to the initial creation of the statefulset without bootstrap env variables.

It's not a hard reqirement for this feature that the spec is populated with the defaults.

Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java:108]: Ok, I see. Thanks for clarifying!

**Pull request merged**
